### ISSUE TITLE
Error in creation of enkf archive file list with DOIAU=YES

### ISSUE THREAD

**Issue by KateFriedman-NOAA** [NOAA-EMC/global-workflow#1623]
**Current behavior**

Generated list of enkf files to archive is wrong. This `enkfgdas.t00z.atmf0036,9.nc` is wrong.

**Machines affected**

All (any that archive to HPSS).

**To Reproduce**

Build and run global-workflow develop with `DOIAU=YES` and enkf archival turned on.

**Additional Information**

Log on Hera: /scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/verfiles/enkfgdasearc01.log.0

Error at bottom of log:
```
++ exgdas_enkf_earc.sh[69]: cat /scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/verfiles/enkfgdas.20220102/00/earc01/enkfgdas_grp1.txt
+ exgdas_enkf_earc.sh[69]: htar -P -cvf /NCEPDEV/emc-global/1year/Kate.Friedman/HERA/scratch/verfiles/2022010200/enkfgdas_grp01.tar enkfgdas.20220102/00/mem001/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem002/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem003/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem004/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem005/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem006/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem007/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem008/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem009/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem010/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc
[connecting to hpsscore1.fairmont.rdhpcs.noaa.gov/1217]
ERROR: /apps/hpss/bin/htar: error -1 stat-ing "enkfgdas.20220102/00/mem001/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc"
*** Warning: No matching names located for 'enkfgdas.20220102/00/mem001/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc'
ERROR: /apps/hpss/bin/htar: error -1 stat-ing "enkfgdas.20220102/00/mem002/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc"
*** Warning: No matching names located for 'enkfgdas.20220102/00/mem002/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc'
ERROR: /apps/hpss/bin/htar: error -1 stat-ing "enkfgdas.20220102/00/mem003/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc"
*** Warning: No matching names located for 'enkfgdas.20220102/00/mem003/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc'
ERROR: /apps/hpss/bin/htar: error -1 stat-ing "enkfgdas.20220102/00/mem004/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc"
*** Warning: No matching names located for 'enkfgdas.20220102/00/mem004/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc'
ERROR: /apps/hpss/bin/htar: error -1 stat-ing "enkfgdas.20220102/00/mem005/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc"
*** Warning: No matching names located for 'enkfgdas.20220102/00/mem005/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc'
ERROR: /apps/hpss/bin/htar: error -1 stat-ing "enkfgdas.20220102/00/mem006/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc"
*** Warning: No matching names located for 'enkfgdas.20220102/00/mem006/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc'
ERROR: /apps/hpss/bin/htar: error -1 stat-ing "enkfgdas.20220102/00/mem007/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc"
*** Warning: No matching names located for 'enkfgdas.20220102/00/mem007/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc'
ERROR: /apps/hpss/bin/htar: error -1 stat-ing "enkfgdas.20220102/00/mem008/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc"
*** Warning: No matching names located for 'enkfgdas.20220102/00/mem008/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc'
ERROR: /apps/hpss/bin/htar: error -1 stat-ing "enkfgdas.20220102/00/mem009/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc"
*** Warning: No matching names located for 'enkfgdas.20220102/00/mem009/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc'
ERROR: /apps/hpss/bin/htar: error -1 stat-ing "enkfgdas.20220102/00/mem010/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc"
*** Warning: No matching names located for 'enkfgdas.20220102/00/mem010/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc'
INFO: refusing to create empty archive
HTAR: HTAR SUCCESSFUL
###WARNING  htar returned non-zero exit status.
            72 = /apps/hpss/bin/htar -P -cvf /NCEPDEV/emc-global/1year/Kate.Friedman/HERA/scratch/verfiles/2022010200/enkfgdas_grp01.tar enkfgdas.20220102/00/mem001/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem002/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem003/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem004/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem005/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem006/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem007/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem008/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem009/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc enkfgdas.20220102/00/mem010/analysis/atmos/enkfgdas.t00z.atmf0036,9.nc
+ exgdas_enkf_earc.sh[1]: postamble exgdas_enkf_earc.sh 1684276225 72
```

Traced list generation back to a list of output files designated by the contents of `IAUFHRS_ENKF`. When `DOIAU=YES` and `DOIAU_ENKF=YES` the value of `IAUFHRS_ENKF` is `3,6,9`.

```
+++ config.base[269]: export DOIAU=YES
+++ config.base[269]: DOIAU=YES
+++ config.base[270]: export IAUFHRS=3,6,9
+++ config.base[270]: IAUFHRS=3,6,9
++++ config.base[271]: echo 3,6,9
++++ config.base[271]: cut -c1
+++ config.base[271]: export IAU_FHROT=3
+++ config.base[271]: IAU_FHROT=3
+++ config.base[272]: export IAU_DELTHRS=6
+++ config.base[272]: IAU_DELTHRS=6
+++ config.base[273]: export IAU_OFFSET=6
+++ config.base[273]: IAU_OFFSET=6
+++ config.base[274]: export DOIAU_ENKF=YES
+++ config.base[274]: DOIAU_ENKF=YES
+++ config.base[275]: export IAUFHRS_ENKF=3,6,9
+++ config.base[275]: IAUFHRS_ENKF=3,6,9
```

Later, while in `hpssarch_gen.sh`...

```
+ hpssarch_gen.sh[472](enkfgdas): IAUFHRS_ENKF=3,6,9
```
...the value of `nfhrs` becomes:
```
+ hpssarch_gen.sh[474](enkfgdas): nfhrs=36,9
```
This should likely be `369` instead of `36,9` or a list with spaces (?) `3 6 9` for a for-loop later.

In `hpssarch_gen.sh`, line 474:
```
468 #-----------------------------------------------------
469 if [[ ${type} == "enkfgdas" || ${type} == "enkfgfs" ]]; then
470 #-----------------------------------------------------
471 
472   IAUFHRS_ENKF=${IAUFHRS_ENKF:-6}
473   lobsdiag_forenkf=${lobsdiag_forenkf:-".false."}
474   nfhrs="${IAUFHRS_ENKF/,/}"
...
501     for FHR in ${nfhrs}; do  # loop over analysis times in window
```
...should probably be:
```
474   nfhrs="${IAUFHRS_ENKF//,}"
```
Change the order of the 2nd forward slash and comma.

Tested this on the command line:
```
-bash-4.2$ IAUFHRS_ENKF="3,6,9"
-bash-4.2$ echo $IAUFHRS_ENKF
3,6,9
-bash-4.2$ echo "${IAUFHRS_ENKF/,/}"
36,9
-bash-4.2$ echo "${IAUFHRS_ENKF//,}"
369
```

**Possible Implementation**

Looking into this further for possible resolution.

**Comment by WalterKolczynski-NOAA** [comment]
That still won't be right, because it needs to be array. It needs to be changed to
```
IFS=',' read -ra nfhrs <<< "${IAUFHRS_ENKF}"
```

**Comment by WalterKolczynski-NOAA** [comment]
Demonstration:
```
$ a='3,6,9'
$ IFS=',' read -ra b <<< "${a}"
$ echo ${b[*]}
3 6 9
$ echo ${b[0]}
3
$ echo ${b[2]}
9
```

**Comment by KateFriedman-NOAA** [comment]
Yep, that appears to work for me when I try it in my test on Hera. Needed to test with fix for #1622 to get successful job completion.

Tested diffs for this and #1622:
```
-bash-4.2$ git diff ush/hpssarch_gen.sh
diff --git a/ush/hpssarch_gen.sh b/ush/hpssarch_gen.sh
index f9a15e23..57eba5b2 100755
--- a/ush/hpssarch_gen.sh
+++ b/ush/hpssarch_gen.sh
@@ -471,7 +471,7 @@ if [[ ${type} == "enkfgdas" || ${type} == "enkfgfs" ]]; then
 
   IAUFHRS_ENKF=${IAUFHRS_ENKF:-6}
   lobsdiag_forenkf=${lobsdiag_forenkf:-".false."}
-  nfhrs="${IAUFHRS_ENKF/,/}"
+  IFS=',' read -ra nfhrs <<< ${IAUFHRS_ENKF}
   NMEM_ENS=${NMEM_ENS:-80}
   NMEM_EARCGRP=${NMEM_EARCGRP:-10}               ##number of ens memebers included in each tarball
   NTARS=$((NMEM_ENS/NMEM_EARCGRP))
@@ -498,7 +498,7 @@ if [[ ${type} == "enkfgdas" || ${type} == "enkfgfs" ]]; then
     if [[ -s "${COM_ATMOS_ANALYSIS_ENSSTAT}/${head}radstat.ensmean" ]]; then
          echo "${COM_ATMOS_ANALYSIS_ENSSTAT/${ROTDIR}\//}/${head}radstat.ensmean"
     fi
-    for FHR in ${nfhrs}; do  # loop over analysis times in window
+    for FHR in ${nfhrs[@]}; do  # loop over analysis times in window
       if [[ ${FHR} -eq 6 ]]; then
         if [[ -s "${COM_ATMOS_ANALYSIS_ENSSTAT}/${head}atmanl.ensmean.nc" ]]; then
           echo "${COM_ATMOS_ANALYSIS_ENSSTAT/${ROTDIR}\//}/${head}atmanl.ensmean.nc"
@@ -560,10 +560,11 @@ if [[ ${type} == "enkfgdas" || ${type} == "enkfgfs" ]]; then
 
       MEMDIR="mem${mem}" YMD=${PDY} HH=${cyc} generate_com \
         COM_ATMOS_ANALYSIS_MEM:COM_ATMOS_ANALYSIS_TMPL \
-        COM_ATMOS_RESTART_MEM:COM_ATMOS_RESTART_TMPL
+        COM_ATMOS_RESTART_MEM:COM_ATMOS_RESTART_TMPL \
+        COM_ATMOS_HISTORY:COM_ATMOS_HISTORY_TMPL
 
       #---
-      for FHR in $nfhrs; do  # loop over analysis times in window
+      for FHR in ${nfhrs[@]}; do  # loop over analysis times in window
         if [ $FHR -eq 6 ]; then
           {
             if (( n <= NTARS2 )); then
@@ -598,9 +599,9 @@ if [[ ${type} == "enkfgdas" || ${type} == "enkfgfs" ]]; then
           fi
         fi
         {
-          echo "${COM_ATMOS_ANALYSIS_MEM/${ROTDIR}\//}/${head}atmf00${FHR}.nc"
+          echo "${COM_ATMOS_HISTORY/${ROTDIR}\//}/${head}atmf00${FHR}.nc"
           if (( FHR == 6 )); then
-            echo "${COM_ATMOS_ANALYSIS_MEM/${ROTDIR}\//}/${head}sfcf00${FHR}.nc"
+            echo "${COM_ATMOS_HISTORY/${ROTDIR}\//}/${head}sfcf00${FHR}.nc"
           fi
         } >> "${RUN}_grp${n}.txt"
       done # loop over FHR
```

Log: /scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/verfiles/logs/2022010200/enkfgdasearc01.log

**Comment by RussTreadon-NOAA** [comment]
@KateFriedman-NOAA and @WalterKolczynski-NOAA , as documented in issue #1454 g-w archive jobs fail when running UFSDA parallels.   UFSDA parallels do not generate the same output as GSI-EnKF based parallels.   It seems that it may be a while before resources are available to generalize our approach to HPSS archive.  Given this @aerorahul said short-term workarounds are acceptable.   

The workaround is simple.  The first step is to add logic in `hpssarch_gen.sh` to check for the existence of GSI-EnKF specific files before adding them to archive lists.   We do this, for example, for cyclone track files via
```
    #Only generated if there are cyclones to track
    cyclone_files=("avno.t${cyc}z.cyclone.trackatcfunix"
                   "avnop.t${cyc}z.cyclone.trackatcfunix"
                   "trak.gfso.atcfunix.${PDY}${cyc}"
                   "trak.gfso.atcfunix.altg.${PDY}${cyc}")

    for file in "${cyclone_files[@]}"; do
      [[ -s ${COM_ATMOS_TRACK}/${file} ]] && echo "${COM_ATMOS_TRACK/${ROTDIR}\//}/${file}"
    done
```

The second step would be to add logic to check for the existence of UFSDA specific files and add them to the archive list if they are present.   These changes are stopgap measures to allow UFSDA cycling with HPSS archival of key analysis output.

Since UFSDA changes will apply to gfs, gdas, and enkf HPSS archives, I'll open a new issue.  I wanted to mention this here since this development will be a follow-on to this PR.   

Issue #1628 opened for the above development.
